File: cpp/128.cpp
// 最长连续序列
// Longest Consecutive Sequence

#include "common.h"
#include <unordered_map>

class Solution
{
public:
    int longestConsecutive(std::vector<int> &nums)
    {
        std::unordered_map<int, bool> num_set;
        for (int num : nums)
        {
            num_set[num] = true;
        }
        int max_len = 0;

        for (auto &[num, _] : num_set)
        {
            if (num_set.find(num - 1) != num_set.end())
            {
                continue;
            }

            int curr_len = 1;
            int curr_num = num;
            while (num_set.find(curr_num + 1) != num_set.end())
            {
                curr_len += 1;
                curr_num += 1;
            }
            max_len = std::max(max_len, curr_len);
        }
        return max_len;
    }
};

int main()
{
    std::vector<std::vector<int>> test_cases = {
        {100, 4, 200, 1, 3, 2},
        {0, 3, 7, 2, 5, 8, 4, 6, 0, 1},
        {1},
        {},
    };

    Solution sol;
    for (auto &nums : test_cases)
    {
        int res = sol.longestConsecutive(nums);
        std::cout << res << std::endl;
    }
}

File: python/128.py
# 最长连续序列
# Longest Consecutive Sequence


class Solution:
    def longestConsecutive(self, nums: list[int]) -> int:
        num_set = set(nums)
        max_len = 0

        for num in num_set:
            if num - 1 in num_set:
                continue

            curr_len = 1
            curr_num = num
            while curr_num + 1 in num_set:
                curr_len += 1
                curr_num += 1

            if curr_len > max_len:
                max_len = curr_len

        return max_len


if __name__ == "__main__":
    test_cases = [
        ([100, 4, 200, 1, 3, 2],),
        ([0, 3, 7, 2, 5, 8, 4, 6, 0, 1],),
        ([1],),
        ([],),
    ]

    sol = Solution()
    for case in test_cases:
        res = sol.longestConsecutive(*case)
        print(case, "\n", res)
